### ISSUE TITLE
Validation agent has datatype train

### ISSUE THREAD

**Issue by Henry-E** [facebookresearch/ParlAI#216]
In examples/train_model.py there is the option to run validation every n seconds during training. However the model agent which observes the teacher's act containing the validation data still has datatype train in its self.opt dictionary. In the example drqa agent whether an act is intended to be used for validation or training is inferred from the [contents of the observation](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/drqa/drqa.py#L197) itself.

In my task it's not possible to distinguish between the validation and training data based on an observation's contents. The simple workaround I'm using is to have a custom act function in a class that inherits from DialogTeacher and pass a validation flag inside the observation dictionary. This seems to go against the spirit of using an opt dictionary but I'm leery of going deeper into the worlds and agent modules.

On a side note having format X-Y-Z for the parse arguments and X_Y_Z for the key dictionaries makes it an initial confusion and minor inconvenience when for example using sublime text's search all files function to find out what the option does or what its default value is.

**Comment by alexholdenmiller** [comment]
That's right--opt['datatype'] is geared towards the data loading, so for the models we were using just the simple heuristic "return a prediction always--if you see a label, then update yourself". I suppose this could be true more generally as "if you receive a reward, update yourself".

If your use case calls for very particular modes in these cases, one option would be to add functions like "agent.train_mode()" or "agent.eval_mode()" to specifically switch these modes--you can modify train_model.py to call these functions before training or evaluating.


As far as the argument names, yes I see what you mean there. Sampling a bunch of different command-line programs, I found that most used hyphens for multiword args, so we did that to match them.

**Comment by Henry-E** [comment]
Training models for dialogue and text generation with a more reinforcement learning approach to  is something I'd like to get into but would have to wrap my head around first. I can see that the framework is probably geared towards nudging research more in that direction than traditional supervised learning paradigms.

Cool, I like the idea of switching between modes. It's kind of like how pytorch does it as well for training and evaluation.

Yeh the argument names isn't really a big issue just thought I'd mention it. Hyphens in the command line arguments does make more sense.

**Comment by Henry-E** [comment]
The dialogue teacher also [removes labels](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/dialog_teacher.py#L136) for non-training datasets, which might not be the best default option. In many language generation tasks correct labels are required for the validation set so that the loss can be calculated.

**Comment by jaseweston** [comment]
The teacher has the labels and can do the evaluation?

**Comment by alexholdenmiller** [comment]
Yep we use the labels during validation to calculate metrics [here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/dialog_teacher.py#L100).

**Comment by Henry-E** [comment]
Ah ok, I didn't realise that's the way it was done till I had to change train_model.py to call agent.report instead of world.report . So far I had thought the idea was to pass only human readable data to and from the model / training agent. To calculate the loss using the labels though would probably require returning a tensor of probabilities and then sending it back to the model again?

**Comment by jaseweston** [comment]
you mean to compute things like perplexity (which we don't currently)..?
to do that yes you'd have to pass probabilities.. you could do that with a
field 'text_candidate_scores' in the message
i find such losses a little weird because you can't evaluate a human with
them...(or a model that doesn't generate probabilities).. but still they
can be useful

**Comment by Henry-E** [comment]
Yes, language generation evaluation metrics do leave a lot to be desired. That's why I'm only using perplexity as a test for overfitting and hyperparameter tuning. I'm getting more into the spirit of ParlAI anyway. It's probably easy enough to expose a loss / perplexity function in the model which can be called via a custom function in metrics.py. Thanks for the advice guys.

**Comment by alexholdenmiller** [comment]
@Henry-E #300 is adding labels to a field `eval_labels`

We're removing `labels` to make sure the agent doesn't accidentally use them during training, but putting them in this field so they're available in case they're useful
